### PR TITLE
[CodeGen][NewPM] Port PeepholeOptimizer to NPM

### PR DESCRIPTION
**Pull request by optimisan**
With this, all machine SSA optimization passes are available in the new codegen pipeline.

### CODE REVIEW

**Comment by optimisan** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116326?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#116326** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116326?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116326?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#116325** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116325?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: llvm/lib/CodeGen/PeepholeOptimizer.cpp [hunk 1]
```diff
@@ -1644,9 +1652,36 @@ bool PeepholeOptimizer::optimizeRecurrence(MachineInstr &PHI) {
...
+      Aggressive ? &MFAM.getResult<MachineDominatorTreeAnalysis>(MF) : nullptr;
+  auto *MLI = &MFAM.getResult<MachineLoopAnalysis>(MF);
+  PeepholeOptimizer Impl(DT, MLI);
+  bool Changed = Impl.run(MF);
+  if (!Changed)
+    return PreservedAnalyses::all();
+
+  auto PA = getMachineFunctionPassPreservedAnalyses();
+  PA.preserve<MachineDominatorTreeAnalysis>();
+  PA.preserve<MachineLoopAnalysis>();
```
Comment (paperchalice) [inline llvm/lib/CodeGen/PeepholeOptimizer.cpp:1669]: Add `PA.preserveSet<CFGAnalyses>();`.

Comment (optimisan) [reply llvm/lib/CodeGen/PeepholeOptimizer.cpp:1669]: somehow missed that, done

**Review by paperchalice — APPROVED** [review]

**Review by arsenm — APPROVED** [review]

**Comment by optimisan** [comment]
### Merge activity

* **Nov 18, 12:24 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/116326).

**Pull request merged**